Feature/eunrain

eunrain opened the pull request:
## 3주차 플렉시스 기업과제 
주어진 데이터를 기반으로 시계열 차트 만들기

https://user-images.githubusercontent.com/113877276/225390585-edd4fe71-c793-4d16-aa59-afdce83c7d96.mov

## 요구사항
1. 시계열 차트 만들기
    - [x] 주어진 JSON 데이터의 `key`값(시간)을 기반으로 시계열 차트를 만들어주세요
    - [x] 하나의 차트안에 Area 형태의 그래프와 Bar 형태의 그래프가 모두 존재하는 복합 그래프로 만들어주세요
    - [x] Area 그래프의 기준값은 `value_area` 값을 이용해주세요
    - [x] Bar 그래프의 기준값은 `value_bar` 값을 이용해주세요
    - [x] 차트의 Y축에 대략적인 수치를 표현해주세요(예시 이미지 참고)
2. 호버 기능 구현
    - [x] 특정 데이터 구역에 마우스 호버시 `id, value_area, value_bar` 데이터를 툴팁 형태로 제공해주세요
3. 필터링 기능 구현
    - [x] 필터링 기능을 구현해주세요, 필터링은 특정 데이터를 하이라이트 하는 방식으로 구현해주세요
    - [x] 필터링 기능은 버튼 형태로 ID값(지역이름)을 이용해주세요
    - [x] 필터링 시 버튼에서 선택한 ID값과 동일한 ID값을 가진 데이터 구역만 하이라이트 처리를 해주세요
    - [x] 특정 데이터 구역을 클릭 시에도 필터링 기능과 동일한 형태로 동일한 ID값을 가진 데이터 구역을 하이라이트해주세요

## 구현 방법
- Recharts 라이브러리 사용
- CustomTooltip 만들어서 호버 기능 구현 
- useParams를 이용하여 필터링 기능 구현 

## 보완해야할 점
 - [ ] Area 그래프 필터링
 - [x] Area 그래프가 bar 그래프 앞으로 올 수 있도록 차트 view 다시 그리기
 - [x] 코드 정리 & 컴포넌트 분리 
 
# 👩🏻‍💻
개인적으로 apexchart로 과제 진행 도중, 필터링 기능 구현이 어려워서 Recharts를 이용해 과제를 마무리하였습니다.
필터링 부분 다들 어떻게 하셨는지 너무 궁금합니다!🥹
2주차 과제 피드백을 반영해서, useparams로 필터링 기능을 구현해보았습니다.

eunrain committed:
[#2] Feat : 차트 구현 완료-recharts

eunrain committed:
[#2] Feat : 호버 기능 구현

-CustomTooltip 을 만들어, 데이터를 툴팁 형태로 제공

eunrain committed:
[#2] Feat : 필터링 기능 구현

whl5105 commented:
저는 chart.js 로 구현했습니다! 
필터링은 드롭다운으로 구현했는데 버튼으로 구현하는것도 좋아보입니다 ! 고생하셨어요 👍

NyeongDev commented:
- 저는 apexcharts로 필터링 기능 구현 할 때 data의 fillColor 속성을 변경해주는 방식으로 구현했습니다! 은비님이 하신 것처럼 useparams로 필터링을 구현하는 것이 페이지 이동 시 또는 url 공유 시 데이터가 유지될 수 있어서 너무 좋은 것 같아요!
- 전체 버튼을 추가해서 사용자 편의성을 고려하신 부분도 좋았습니다!

수고하셨습니다~👍

minhyeonhong commented:
저도 준수님 코드보고 은비님처럼 data의 fillColor속성을 변경하는 방식으로 하이라이트 구현했습니다.
필터링 정보를 2주차 과제 피드백을 반영해서, useparams로 필터링 기능구현 하신거 너무 멋지십니다.
고생하셨습니다.

Hwang-Junsu commented:
useParam을 사용해서 필터링 데이터를 유지하는 것이 좋은 것 같습니다! 
저도 fillColor 속성을 사용해서 하이라이트를 구현했습니다.
고생하셨습니다!

eunrain committed:
[#2] Refactor : 컴포넌트 분리 및 코드 정리

- useGetData hook을 만들어, 데이터를 가져오고 가공하는 부분을 따로 빼서 관리
- tsconfig.json에서 target을 es6로 바꾸어, new Set 문법 적용

eunrain committed:
[#2] Refactor : Area 그래프 view 수정

- fillOpacity를 이용해서 투명도를 낮추어, area그래프가 보이도록 만들어 주었습니다

eunrain committed:
[#2] Refactor : useSearchParams 적용해서 필터링 구현

eunrain committed:
[#2] Refactor : 코드 정리 및 안쓰는 파일 제거

eunrain commented:
## Refactor/eunrain
https://user-images.githubusercontent.com/113877276/225708242-dbaed7f3-3596-45b0-b41a-3dbaeb08aa70.mov

## 😎 개선 사항 
- `useGetData` 로 관심사 분리, 데이터를 불러오고 가공하는 부분의 유지보수가 용이하도록 구현했습니다.
- `useSearchParams` 를 이용해서 필터링 기능을 구현했습니다.
-  tsconfig.json에서 target을 es6로 바꾸어, new Set 문법을 적용해 코드를 간결하게 작성했습니다.
- fillOpacity를 이용해서 투명도를 낮추어, area의 그래프가 보여지도록 만들었습니다.
- 🚫 area 필터링은 구현하지 못했습니다...🚫

## 🙋🏻‍♀️ 궁금한 점 
첫번째 PR 때 useparams를 사용해서 필터링 기능을 구현했습니다. 
리팩토링을 진행하며 useSearchParams를 사용해서 필터링을 다시 구현했는데, 이전과 코드가 많이 달라진 것 같지 않아서 이게 맞나.. 싶으면서 조금 의아했습니다..
그래서 둘의 차이점을 찾다보니, useSearchParams는 페이지 라우팅 경로에 영향을 주지 않는다는 점에서 useparams와 약간 다르다고 생각하는데, 비슷한 거 같기도 하고 사실.. 아직까지도 잘 모르겠습니다 🥹 
이 부분에 대해서 같이 얘기 나눠보면 좋을 것 같습니다...!

yeon-kk commented on src/apis/instance.ts:
```diff
@@ -2,11 +2,13 @@ import axios from 'axios';
 
-export default instance;
+export const DataAPI = {
+  getData: () => instance.get(''),
```
혹시 함수를 호출할 때 의도하신 함수명이 있으셔서 getData를 key로 하는 객체를 전달하셨을지 궁금합니다

eunrain replied:
의도한 바는 따로 없습니다..😅 호출할 함수가 getData 하나여서 연경님이 말씀하신 대로 처리해주는게 훨씬 깔끔할 것 같습니다!

yeon-kk commented:
수인님이 사용하신 QueryParmRoute는 hooks폴더의 useParams파일에서 useSearchParamsState는 custom hooks처럼 사용하신 것 같아요. 
searchParam이 null/undefined이면 defaultValue인 ''를 할당하신 것 같습니다. 
QueryParmRoute에서 사용하신 useSearchParamsState()는 searchParamsState만 반환해서 if문으로 require && searchParamsState==='' 조건과 비교해서 분기하신 것 같아요

NyeongDev commented:
searchParams는 해당 키워드가 어떤 키워드를 검색하거나, 요청 할 때 필요한 '옵션'임을 명시적으로 전달하고 싶을 때 사용하는 것 같아요!

라우팅 경로에 영향을 주지 않는다는 점은 저도 잘 모르겠습니다.. 혹시 라우터 엘리먼트가 바뀌지 않는다는 뜻은 아닐까요?

eunrain committed:
[#2] Refactor : instance 코드 정리

eunrain commented:
📍 useParams 와 useSearchParams의 차이 

useParams와 useSearchParams은 React Router의 훅으로, 둘 다 URL 파라미터를 가져오는 데 사용됩니다. 그러나 둘 사이에는 중요한 차이점이 있습니다.

useParams는 동적 라우팅을 위해 사용됩니다. 즉, 라우트 경로에서 매개변수가 동적으로 변경될 때 사용됩니다. 예를 들어, /users/:id 라우트에서 id 매개변수를 가져와서 해당 사용자의 정보를 가져오는 데 사용할 수 있습니다. useParams는 이러한 매개변수를 가져와서 컴포넌트에서 사용할 수 있도록 해줍니다.

반면에 useSearchParams는 쿼리 문자열에서 매개변수를 가져옵니다. 이는 검색 기능과 관련이 있습니다. 예를 들어, /?q=react&page=2 URL에서 q와 page 매개변수를 가져올 수 있습니다. useSearchParams는 이러한 매개변수를 가져와서 컴포넌트에서 사용할 수 있도록 해줍니다.

사용 목적에 따라서 알맞게 선택해서 사용하면 될 것 같습니다..!

예를들어 
Path Parameter(한가지 정보만) - 상세페이지 이동할때
Query Parameter(정보가 많을때) - 페이지네이션, 필터링
